Source/AwsCommonRuntimeKit/crt/CStruct.swift

```diff
@@ -0,0 +1,41 @@
+//  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//  SPDX-License-Identifier: Apache-2.0.
+import AwsCIo
+import AwsCCommon
+// This file defines the protocols & helper functions for C Structs.
+// Instances implementing this protocol should define RawType as their C Struct.
+protocol CStruct<RawType> {
+    associatedtype RawType
+    func withCStruct<Result>(shutdownOptions: aws_shutdown_callback_options?, _ body: (RawType) -> Result) -> Result
+}
+
+extension CStruct {
+    func withCStruct<Result>( _ body: (RawType) -> Result) -> Result {
+        withCStruct(shutdownOptions: nil, body)
+    }
+
+    func withCPointer<Result>(shutdownOptions: aws_shutdown_callback_options?, _ body: (UnsafePointer<RawType>) -> Result) -> Result {
+        return withCStruct(shutdownOptions: shutdownOptions) { cStruct in
+            return withUnsafePointer(to: cStruct) { body($0) }
+        }
+    }
+
+    func withCPointer<Result>(_ body: (UnsafePointer<RawType>) -> Result) -> Result {
+        return withCStruct { cStruct in
+            return withUnsafePointer(to: cStruct) { body($0) }
+        }
+    }
+}
+
+func withOptionalCStructPointer<T, Result>(
+        to arg1: (any CStruct)?, _ body: (UnsafePointer<T>?) -> Result
+) -> Result {
+    if let arg1 = arg1 {
+        return arg1.withCStruct { cStruct in
+            return withUnsafePointer(to: cStruct as! T) { structPointer in
+                body(structPointer)
+            }
+        }
+    }
+    return body(nil)
+}
```

Comment:
We will need to define more overrides of this functions as T1, T2, T3.... I tried to find a way to make it generic by passing an array or something but could not get it to work.